File: tests/journal.lib/memo_sanity.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>


// type aliases
using ansi_t = pyre::journal::ansi_t;
using memo_t = pyre::journal::memo_t;
using debug_t = pyre::journal::debug_t;
using chronicler_t = pyre::journal::chronicler_t;


// exercise the {memo} renderer
int main() {
    // grab the global metadata table from chronicler
    auto & globals = chronicler_t::notes();
    // set some metadata
    globals["application"] = "memo";
    globals["author"] = "michael";

    // use a {debug_t} to build a document and its metadata
    debug_t channel("memo");
    // put some stuff in it; careful not to flush so we don't lose everything
    channel
        << pyre::journal::at(__HERE__)
        << pyre::journal::note("time", "now")
        << pyre::journal::note("device", "null")
        << "simon says:"
        << pyre::journal::newline
        << "    hello world!"
        << pyre::journal::newline;

    // make a palette
    memo_t::palette_type palette;
    // add some decorations
    palette["reset"] = ansi_t::x11("normal");
    palette["channel"] = ansi_t::x11("light slate gray");
    palette["debug"] = ansi_t::x11("steel blue");
    palette["body"] = "";

    // pull the contents
    auto & entry = channel.entry();
    // make a memo
    memo_t memo;
    // ask it to render what we have
    memo.render(palette, entry);

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/memo_sanity.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Exercise the developer facing renderer
    """
    # get the renderer
    from journal.Memo import Memo as memo
    # the color spaces
    from journal.ANSI import ANSI
    # and a channel
    from journal.Debug import Debug as debug

    # get the chronicler metadata
    gmeta = debug.chronicler.notes
    # add some
    gmeta["application"] = "memo"
    gmeta["author"] = "michael"

    # make a channel
    channel = debug(name="tests.journal.debug")
    # generate a fake stack trace
    channel.notes["filename"] = "memo_sanity"
    channel.notes["line"] = 29
    channel.notes["function"] = "test"
    # add some metadata
    channel.notes["time"] = "now"
    channel.notes["device"] = "null"
    # inject
    channel.line("debug channel:")
    channel.line("    hello world!")

    # make a palette
    palette = {
        "reset": ANSI.x11("normal"),
        "channel": ANSI.x11("light slate gray"),
        "debug": ANSI.x11("steel blue"),
        "body": "",
        }

    # instantiate the renderer
    renderer = memo()
    # ask it to do its thing
    page = '\n'.join(renderer.render(palette=palette, entry=channel.entry))
    # and show me
    # print(page)

    # all done
    return


# main
if __name__ == "__main__":
    # prohibit the journal bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file
